drivers: uart_gecko: added support for hardware flow control

tidamm opened the pull request:
Added support for hardware flow control in EFM32 UART driver.

Tested with samples/hello_world on efm32pg_stk3402a and further low level tests with uart_poll_in() and uart_poll_out() as in tests/drivers/uart/uart_basic_api/.

Signed-off-by: Timo Dammes <timo.dammes@lemonbeat.com>

tidamm committed:
drivers: uart_gecko: added support for hardware flow control

Added support for hardware flow control in EFM32 UART driver.

Signed-off-by: Timo Dammes <timo.dammes@lemonbeat.com>
Signed-off-by: Christian Taedcke <christian.taedcke@lemonbeat.com>

tidamm commented:
@galak Hello, can you provide some feedback what is missing here and prevents this from being merged? BR

mnkp commented on drivers/serial/uart_gecko.c:
```diff
@@ -301,33 +432,105 @@ static const struct uart_driver_api uart_gecko_driver_api = {
+ * @param uart_type UART or USART
+ * @param nr number of UART/USART (0,...,3 for USART0,...,USART3)
+ */
+#define UART_PINS_WITH_FLOW_CONTROL(uart_type, nr) \
+	.pin_rx  = UART_PIN_INIT(uart_type, nr, RX, gpioModeInput), \
+	.pin_tx  = UART_PIN_INIT(uart_type, nr, TX, gpioModePushPull), \
+	.pin_rts = UART_PIN_INIT(uart_type, nr, RTS, gpioModePushPull), \
+	.pin_cts = UART_PIN_INIT(uart_type, nr, CTS, gpioModeInput),
+
+/**
+ * @brief Set GPIO assignment for UART pins (RX, TX)
+ *
+ * @param uart_type UART or USART
+ * @param nr number of UART/USART (0,...,3 for USART0,...,USART3)
+ */
+#define UART_PINS(uart_type, nr) \
+	.pin_rx  = UART_PIN_INIT(uart_type, nr, RX, gpioModeInput), \
+	.pin_tx  = UART_PIN_INIT(uart_type, nr, TX, gpioModePushPull),
```
I believe the driver instantiation code would be simpler if we defined two macros `UART_RX_TX_PINS` and `UART_CTS_RTS_PINS` and then included the latter only if HW flow control is enabled for the instance. It should also make it easier to update the driver in the future to use macro based driver instantiation, similar to how other drivers, e.g. gpio_gecko.c are doing. Though in case of the uart driver this is not going to be straightforward.

mnkp commented on drivers/serial/uart_gecko.c:
```diff
@@ -232,26 +305,79 @@ static void uart_gecko_isr(void *arg)
+
+#ifdef HW_FLOWCONTROL_IS_SUPPORTED_BY_SOC
+	/* configure HW flow control (RTS, CTS) */
+	if (config->hw_flowcontrol != usartHwFlowControlNone) {
```
It would cleaner to write it as
```
	if (config->hw_flowcontrol == usartHwFlowControlCtsAndRts) {
```

chrta commented:
I am taking over with this pr from my colleague. This change is based on #25977 (which is just one commit). So i can either remove #25977 or we wait with merging this until #25977 is merged.

All comments should be addressed.

chrta committed:
drivers: serial: gecko: Replace custom macros with _CONCAT

Remove the macros CLOCK_ID_PRFX2 and CLOCK_ID_PRFX and replace them with
_CONCAT from common.h.

Signed-off-by: Christian Taedcke <christian.taedcke@lemonbeat.com>

mnkp commented on drivers/serial/uart_gecko.c:
```diff
@@ -354,19 +498,77 @@ static const struct uart_driver_api uart_gecko_driver_api = {
+#endif /* CONFIG_SOC_GECKO_HAS_INDIVIDUAL_PIN_LOCATION */
+
+#define PIN_UART_RTS(idx)						       \
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(idx, location_rts),		       \
+	({								       \
+		DT_INST_PROP_BY_IDX(idx, location_rts, 1),		       \
+			DT_INST_PROP_BY_IDX(idx, location_rts, 2),	       \
+			gpioModePushPull, 1				       \
+	}),								       \
+	({}))
```
I believe we can't do this. The structure will be initialized to all 0s. The side effect of running `soc_gpio_configure` with such data will be to disable pin PA0. The code was probably supposed not to configure RTS pin at all but that's not the case. It may be easier to force user to provide location property for RTS, CTS pins. Otherwise implementation becomes complicated.

chrta replied:
I had to leave this COND_CODE_1 macros to be able to compile the driver  when one post has enabled hw flow control, and one other port is used without hw flow control.

The rts and cts pins are only configured, if hw flow control is enabled:
```
if (config->hw_flowcontrol) {
		soc_gpio_configure(&config->pin_rts);
		soc_gpio_configure(&config->pin_cts);
```

mnkp replied:
My bad, I overlooked this. Thanks for the comment.

mnkp commented on drivers/serial/uart_gecko.c:
```diff
@@ -13,23 +14,84 @@
 	CMU_Clock_TypeDef clock;
 	uint32_t baud_rate;
+#ifdef UART_GECKO_HW_FLOW_CONTROL
+	USART_HwFlowControl_TypeDef hw_flowcontrol;
```
I think the code would look cleaner if we had `bool` here. Then only on the line `usartInit.hwFlowControl = ` where we write to the usart init struct we should use the proper HAL type. The `MAP_HW_FLOWCONTROL_INIT_VAL` macro could be removed.

chrta replied:
done

chrta commented on drivers/serial/uart_gecko.c:
```diff
@@ -42,6 +104,17 @@ struct uart_gecko_data {
+ * @param dev The device to read from
+ * @param c Pointer to received data
+ */
 static int uart_gecko_poll_in(struct device *dev, unsigned char *c)
```
Removed the redundant doc, also some doc that i think is not necessary.

chrta commented on drivers/serial/uart_gecko.c:
```diff
@@ -232,26 +322,74 @@ static void uart_gecko_isr(void *arg)
+		/*
+		 * set enable flag in USARTn_ROUTEPEN
+		 * (I/O Routing Pin Enable Register)
+		 */
```
done

chrta commented on drivers/serial/uart_gecko.c:
```diff
@@ -232,26 +322,74 @@ static void uart_gecko_isr(void *arg)
+		/*
+		 * set location of RTS/CTS in USARTn_ROUTELOC1
+		 * (I/O Routing Location Register)
+		 */
```
done

chrta commented on drivers/serial/uart_gecko.c:
```diff
@@ -354,19 +498,77 @@ static const struct uart_driver_api uart_gecko_driver_api = {
+#define VALIDATE_GECKO_UART_RTS_CTS_PIN_LOCATIONS(idx)			       \
+	BUILD_ASSERT(DT_INST_PROP_BY_IDX(idx, location_rts, 0) ==	       \
+		      DT_INST_PROP_BY_IDX(idx, location_cts, 0),	       \
+		     "DTS location-* properties must have identical value")
```
This check is now removed. I added a check at the top of the file:
```
#if defined(UART_GECKO_HW_FLOW_CONTROL) && \
	(!defined(CONFIG_SOC_GECKO_HAS_INDIVIDUAL_PIN_LOCATION))
#error "Driver not supporting hardware flow control for this SOC"
#endif
```

chrta commented on drivers/serial/uart_gecko.c:
```diff
@@ -354,19 +449,81 @@ static const struct uart_driver_api uart_gecko_driver_api = {
+		   (0)),
+#define VALIDATE_GECKO_UART_RTS_CTS_PIN_LOCATIONS(idx)			       \
+	COND_CODE_1(DT_INST_PROP(idx, hw_flow_control),			       \
+		    (BUILD_ASSERT(DT_INST_NODE_HAS_PROP(idx, location_rts) &&  \
```
This ensures that location-rts and location-cts are set in the dts, when hw-flow-control is enabled.
But there are a lot of errors displayed in that case. You have to scroll up to the first error to the the message "DTS location-rts and location-cts are mandatory".
Maybe there is a better way to do this with a cleaner error message.

galak replied:
There isn't a better solution available today.  We'd need to look at adding some additional syntax to the binding to convey if prop X than require prop Y.

chrta replied:
Converted this into an issue #27483

mnkp commented on drivers/serial/uart_gecko.c:
```diff
@@ -354,19 +449,81 @@ static const struct uart_driver_api uart_gecko_driver_api = {
+			DT_INST_PROP_BY_IDX(idx, location_rts, 2),	       \
+			gpioModePushPull, 1				       \
+	}),								       \
+	({}))
```
`{}` is a GNU extension. Since we're planning to support other toolchains it would be better to change it to the standard `{0}`.

chrta replied:
done

mnkp reviewed (commented):
I overlooked one minor issue in my last review.

chrta commented:
@ioannisg I prefer not to review/approve code i have partly written myself. @mnkp already approved. So one review/approval from another person is missing.

ioannisg commented:
> @ioannisg I prefer not to review/approve code i have partly written myself. @mnkp already approved. So one review/approval from another person is missing.

Right, Just realized that, thanks.

ioannisg commented:
@mnkp assigning this PR to you.
You might have an idea of a second reviewer here.

carlescufi commented:
@Oanerer and @l-alfred would you mind reviewing?

chrta commented:
@mnkp I believe your approval was removed by accident, can you please re-approve?

LeGrosFaisan reviewed (approved):
Everything seems fine to me. I will probably add these changes to the Gecko Series 2 PR #27042 once finalised so that it also supports HW flow control.

mnkp commented:
> @mnkp assigning this PR to you.
> You might have an idea of a second reviewer here.

@galak could have a quick look and verify usage of DTS access macros. Otherwise, this PR LGTM.

MaureenHelm commented on drivers/serial/uart_gecko.c:
```diff
@@ -18,18 +19,75 @@
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT silabs_gecko_usart
+
+#if (DT_INST_PROP(0, hw_flow_control) ||    \
+	DT_INST_PROP(1, hw_flow_control) || \
+	DT_INST_PROP(2, hw_flow_control) || \
+	DT_INST_PROP(3, hw_flow_control) || \
+	DT_INST_PROP(4, hw_flow_control) || \
+	DT_INST_PROP(5, hw_flow_control))
```
@galak is there a macro we can use to detect if this property exists on any instance? I worry that this will break if there is ever a 6th instance

galak replied:
We dont have a macro right now.  Should see if this can be written using `DT_INST_FOREACH_STATUS_OKAY()`

@mbolivar-nordic any thoughts here?

mbolivar-nordic replied:
> @mbolivar-nordic any thoughts here?

@galak @MaureenHelm  yep, you can use `DT_INST_FOREACH_STATUS_OKAY()` for this as long as you're actually asking the question about status okay instances. The test as written is about any instance at all, not just status okay instances, so it's a bit different. (That said, I suspect the code as written might be wrong since I'm guessing disabled instances should not be taken into account, but that's just a hunch.)

The rest of this comment applies to the "status okay" case, but it would be trivial to add `DT_INST_FOREACH()` to devicetree.h if you really need the result for *all* instances, including disabled ones.

Any question "is `MY_TEST(inst)` true for any status okay `DT_DRV_COMPAT` instance number `inst`?" can be written as something like a [fold](https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29) using DT_INST_FOREACH_STATUS_OKAY.

A code template is:

```
#define MY_FOLD_OP(inst) MY_TEST(inst) ||
#define MY_FOLD_RESULT   DT_INST_FOREACH_STATUS_OKAY(MY_FOLD_OP) 0
```

When expanded:

```
MY_FOLD_RESULT ->
MY_FOLD_OP(0) MY_FOLD_OP(1) ... MY_FOLD_OP(n-1) 0 ->
MY_TEST(0) || MY_TEST(1) || ... MY_TEST(n-1) || 0
```

So as long as `MY_TEST` expands to an expression the preprocessor can evaluate, the entire `MY_FOLD_RESULT` does too.

In this case:

```
#define HAS_HFC_OR(inst) DT_INST_NODE_HAS_PROP(inst, hw_flow_control) ||
#define ANY_INST_HAS_HFC DT_INST_FOREACH_STATUS_OKAY(HAS_HFC_OR) 0

#if ANY_INST_HAS_HFC

...

#endif
```

@galak I will leave it up to you to decide if we should add something like `DT_ANY_INST_STATUS_OKAY(MY_TEST)` to devicetree.h based on the above.

chrta replied:
Thank you, that is a good improvement. I included your suggestion.

MaureenHelm reviewed (commented):
@galak please have a look at the dts macro usages

mbolivar-nordic commented on drivers/serial/uart_gecko.c:
```diff
@@ -13,23 +14,74 @@
+#define HW_FLOWCONTROL_IS_SUPPORTED_BY_SOC
+#endif
+
+#define HAS_HFC_OR(inst) DT_INST_NODE_HAS_PROP(inst, hw_flow_control) ||
```
I missed that this was a boolean property when I first responded. Boolean properties should be tested with DT_INST_PROP() or DT_PROP(), not ...NODE_HAS_PROP(). The bindings type system knows that a boolean property is true if present and false if absent.
https://docs.zephyrproject.org/latest/guides/dts/api-usage.html#property-access

chrta replied:
done

mbolivar-nordic reviewed (commented):
DT macros appear to have been addressed as discussed.

MaureenHelm commented:
@chrta please have a look at the buildkite failures

chrta commented:
@MaureenHelm I just did. There is still an issue:
```
$ cd /tests/kernel/common
$ west build -b efm32hg_slstk3400a --pristine

....

SRAM:        8312 B         8 KB    101.46%
/opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/../../../../arm-zephyr-eabi/bin/ld: region `SRAM' overflowed by 120 bytes
collect2: error: ld returned 1 exit status
```

This also fails for me on the current master. This SoC only has 8kByte RAM, maybe i have to reduce some thread stack sizes. I remember that was necessary in another test, also for some STM32 with only 8kB RAM.

MaureenHelm commented:
> This also fails for me on the current master. This SoC only has 8kByte RAM, maybe i have to reduce some thread stack sizes. I remember that was necessary in another test, also for some STM32 with only 8kB RAM.

Try adding `min_ram` to `tests/kernel/common/testcase.yaml`

chrta commented:
I opened https://github.com/zephyrproject-rtos/zephyr/issues/27583 since sanitycheck did not fail when i execute:
```
$ sanitycheck -p efm32hg_slstk3400a
```
But the build logs indicate the same error i pasted above.

chrta commented:
The SRAM overflow is also present for other SoCs and in other tests and intentionally ignored by sanitycheck, so no change here necessary regarding min_ram in the testcase.yaml.
